--- ungrab-winmodem.c~	2013-02-21 01:59:47.555599323 +0100
+++ ungrab-winmodem.c	2013-02-21 02:01:03.650289618 +0100
@@ -41,7 +41,11 @@
 #define PCI_VENDOR_ID_SLI1543                   0x1543 /* Silicon Inst 1543:3052 sub 1543:3000 */
 
 
-static struct pci_device_id softmodem_pci_tbl [] __devinitdata = {
+static struct pci_device_id softmodem_pci_tbl []
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinitdata
+#endif
+ = {
 	/* 1543:3052 */
 	{PCI_VENDOR_ID_SLI1543, PCI_DEVICE_ID_SL1900, PCI_ANY_ID, PCI_ANY_ID },
 	/* 1057:3052 */
