--- r8168-8.031.00/src/r8168_n.c~	2012-05-17 08:09:42.000000000 +0200
+++ r8168-8.031.00/src/r8168_n.c	2013-02-21 08:46:17.765418157 +0100
@@ -14200,7 +14200,10 @@ rtl8168_phy_power_down (struct net_devic
 	spin_unlock_irqrestore(&tp->phy_lock, flags);
 }
 
-static int __devinit
+static int
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinit
+#endif
 rtl8168_init_board(struct pci_dev *pdev,
 		   struct net_device **dev_out,
 		   void __iomem **ioaddr_out)
@@ -14364,7 +14367,10 @@ err_out:
 	goto out;
 }
 
-static void __devinit
+static void 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinit
+#endif
 rtl8168_init_sequence(struct rtl8168_private *tp)
 {
 	void __iomem *ioaddr = tp->mmio_addr;
@@ -14577,7 +14583,10 @@ static const struct net_device_ops rtl81
 };
 #endif
 
-static int __devinit
+static int
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinit
+#endif
 rtl8168_init_one(struct pci_dev *pdev,
 		 const struct pci_device_id *ent)
 {
@@ -14742,7 +14751,10 @@ rtl8168_init_one(struct pci_dev *pdev,
 	return 0;
 }
 
-static void __devexit
+static void
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devexit
+#endif
 rtl8168_remove_one(struct pci_dev *pdev)
 {
 	struct net_device *dev = pci_get_drvdata(pdev);
@@ -17238,7 +17250,11 @@ static struct pci_driver rtl8168_pci_dri
 	.name		= MODULENAME,
 	.id_table	= rtl8168_pci_tbl,
 	.probe		= rtl8168_init_one,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
 	.remove		= __devexit_p(rtl8168_remove_one),
+#else
+	.remove		= rtl8168_remove_one,
+#endif
 #if LINUX_VERSION_CODE > KERNEL_VERSION(2,6,11)
 	.shutdown	= rtl8168_shutdown,
 #endif
