--- DPO_RT3562_3592_3062_LinuxSTA_V2.4.1.1_20101217/os/linux/pci_main_dev.c~	2013-02-21 00:06:35.599402492 +0100
+++ DPO_RT3562_3592_3062_LinuxSTA_V2.4.1.1_20101217/os/linux/pci_main_dev.c	2013-02-21 08:57:03.437036542 +0100
@@ -34,8 +34,16 @@
 extern int rt28xx_close(IN struct net_device *net_dev);
 extern int rt28xx_open(struct net_device *net_dev);
 
-static VOID __devexit rt2860_remove_one(struct pci_dev *pci_dev);
-static INT __devinit rt2860_probe(struct pci_dev *pci_dev, const struct pci_device_id  *ent);
+static VOID
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devexit
+#endif
+rt2860_remove_one(struct pci_dev *pci_dev);
+static INT
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinit
+#endif
+rt2860_probe(struct pci_dev *pci_dev, const struct pci_device_id  *ent);
 static void __exit rt2860_cleanup_module(void);
 static int __init rt2860_init_module(void);
 
@@ -57,7 +65,11 @@ static int rt2860_resume(struct pci_dev
 //
 // Ralink PCI device table, include all supported chipsets
 //
-static struct pci_device_id rt2860_pci_tbl[] __devinitdata =
+static struct pci_device_id rt2860_pci_tbl[]
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinitdata
+#endif
+ =
 {
 #ifdef RT2860
 	{PCI_DEVICE(NIC_PCI_VENDOR_ID, NIC2860_PCI_DEVICE_ID)},		//RT28602.4G
@@ -101,7 +113,9 @@ static struct pci_driver rt2860_driver =
     name:       "rt3562",
     id_table:   rt2860_pci_tbl,
     probe:      rt2860_probe,
-#if LINUX_VERSION_CODE >= 0x20412
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(3, 8, 0)
+    remove:     rt2860_remove_one,
+#elif LINUX_VERSION_CODE >= 0x20412
     remove:     __devexit_p(rt2860_remove_one),
 #else
     remove:     __devexit(rt2860_remove_one),
@@ -295,7 +309,11 @@ module_exit(rt2860_cleanup_module);
 //
 // PCI device probe & initialization function
 //
-static INT __devinit   rt2860_probe(
+static INT
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinit
+#endif
+rt2860_probe(
     IN  struct pci_dev              *pci_dev, 
     IN  const struct pci_device_id  *pci_id)
 {
@@ -469,7 +487,11 @@ err_out:
 }
 
 
-static VOID __devexit rt2860_remove_one(
+static VOID
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devexit
+#endif
+rt2860_remove_one(
     IN  struct pci_dev  *pci_dev)
 {
 	PNET_DEV	net_dev = pci_get_drvdata(pci_dev);
