*** dotpac.orig	2006-10-13 17:28:22.000000000 +0200
--- dotpac	2012-09-17 22:21:49.492889480 +0200
***************
*** 7,13 ****
  	exit 4
  fi
  
! for e in "$EDITOR" "$VISUAL" vim vi joe mcedit nano pico e3 jed ne emacs; do
  	w=`which $e 2> /dev/null`
  	if [ -n "$w" -a -x "$w" ]; then
  		texted="$w"
--- 7,13 ----
  	exit 4
  fi
  
! for e in "$EDITOR" "$VISUAL" nano pico vim vi joe mcedit e3 jed ne emacs; do
  	w=`which $e 2> /dev/null`
  	if [ -n "$w" -a -x "$w" ]; then
  		texted="$w"
***************
*** 89,95 ****
  
  function menu_actions()
  {
! 	diff -u "$1" "$2" > $tmpdir/diff
  
  	if [ -s $tmpdir/diff ]; then
  		menu_common 3 1
--- 89,95 ----
  
  function menu_actions()
  {
! 	diff -u "$2" "$1" > $tmpdir/diff
  
  	if [ -s $tmpdir/diff ]; then
  		menu_common 3 1
***************
*** 219,224 ****
--- 219,235 ----
  	fi
  }
  
+ 
+ menu_common 9 3
+ cat << EOF >> $tmpdir/dialog
+ --ok-label "Next"
+ --msgbox "You should be aware about what you are doing! Wrong pacnew merging could break your system!"
+ 8 74
+ EOF
+ 
+ dialog --file $tmpdir/dialog
+ 
+ 
  bye=0
  
  while [ $bye == 0 ]; do
