--- fritz/src/main.c~	2013-02-21 00:01:20.364094146 +0100
+++ fritz/src/main.c	2013-02-21 01:00:40.942152010 +0100
@@ -189,7 +189,11 @@ static void isa_stop (void) {
 \*---------------------------------------------------------------------------*/
 #if defined (__fcpnp__)
 
-static int __devinit fritz_probe (
+static int
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinit
+#endif
+fritz_probe (
 	struct pnp_dev *		dev,
 	const struct pnp_device_id *	id
 ) {
@@ -227,7 +231,11 @@ static int __devinit fritz_probe (
 \*---------------------------------------------------------------------------*/
 #if defined (__fcpnp__)
 
-static void __devexit fritz_remove (struct pnp_dev * dev) {
+static void
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devexit
+#endif
+fritz_remove (struct pnp_dev * dev) {
 	card_t *	card;
 	
 	card = (card_t *) pnp_get_drvdata (dev);
@@ -259,7 +259,11 @@ fritz_remove (struct pnp_dev * dev) {
 \*---------------------------------------------------------------------------*/
 #if defined (__fcpci__)
 
-static int __devinit fritz_probe (
+static int
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinit
+#endif
+fritz_probe (
 	struct pci_dev *		dev,
 	const struct pci_device_id *	id
 ) {
@@ -292,7 +300,11 @@ static int __devinit fritz_probe (
 \*---------------------------------------------------------------------------*/
 #if defined (__fcpci__)
 
-static void __devexit fritz_remove (struct pci_dev * dev) {
+static void
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devexit
+#endif
+fritz_remove (struct pci_dev * dev) {
 	card_t *	card;
 	
 	card = (card_t *) pci_get_drvdata (dev);
@@ -320,7 +332,11 @@ static struct pci_driver	fcpci_driver =
 	.name		= TARGET,
 	.id_table	= fcpci_id_table,
 	.probe		= fritz_probe,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
 	.remove		= __devexit_p(fritz_remove),
+#else
+	.remove		= fritz_remove,
+#endif
 } ;
 
 #elif defined (__fcpnp__)
@@ -330,7 +346,11 @@ static struct pnp_driver	fcpnp_driver =
 	.name		= TARGET,
 	.id_table	= fcpnp_id_table,
 	.probe		= fritz_probe,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
 	.remove		= __devexit_p(fritz_remove),
+#else
+	.remove		= fritz_remove,
+#endif
 } ;
 
 #endif
