--- drivers/amrmo_init.c~	2013-02-21 02:08:38.608379071 +0100
+++ drivers/amrmo_init.c	2013-02-21 02:10:30.837792136 +0100
@@ -240,7 +240,11 @@ static const char *card_names[] = {
 };
 
 
-static struct pci_device_id amrmo_pci_tbl [] __devinitdata = {
+static struct pci_device_id amrmo_pci_tbl []
+#if LINUX_VERSION_CODE < KERNEL_VERSION(3, 8, 0)
+__devinitdata
+#endif
+ = {
 	{PCI_VENDOR_ID_SL1543, PCI_DEVICE_ID_SL1900,
 	 PCI_ANY_ID, PCI_ANY_ID, 0, 0, SL1900_CARD},       /* 1543:3052 , Motorola 1543:3000 */
 	{PCI_VENDOR_ID_MOTOROLA, PCI_DEVICE_ID_SL1900,     /* Motorola 1057:3052 */
