--- 12.6/common/lib/modules/fglrx/build_mod/kcl_ioctl.c	2012-05-29 20:18:09.000000000 +0000
+++ 12.6/common/lib/modules/fglrx/build_mod/kcl_ioctl.c	2012-06-01 20:06:35.338329440 +0000
@@ -217,6 +217,9 @@ void ATI_API_CALL KCL_IOCTL_UnregisterCo
  *  \param size [in] Number of bytes to allocate
  *  \return Pointer to allocated memory
  */
+
+DEFINE_PER_CPU(unsigned long, old_rsp);
+
 void* ATI_API_CALL KCL_IOCTL_AllocUserSpace32(long size)
 {
     void __user *ret = COMPAT_ALLOC_USER_SPACE(size);
--- 12.6/common/lib/modules/fglrx/build_mod/firegl_public.c	2012-05-29 20:18:09.000000000 +0000
+++ 12.6/common/lib/modules/fglrx/build_mod/firegl_public.c	2012-06-01 20:08:06.891305714 +0000
@@ -4156,7 +4156,7 @@ static int kasInitExecutionLevels(unsign
 {
     unsigned int p;
     KCL_DEBUG5(FN_FIREGL_KAS, "%d\n", level_init);
-    for_each_cpu_mask(p, cpu_possible_map)
+    for_each_possible_cpu(p)
     {
         KCL_DEBUG1(FN_FIREGL_KAS,"Setting initial execution level for CPU # %d\n", p);
         preempt_disable();
